File: apps/api-py/eq_api/routes/iam.py
"""
Organizaciones, miembros, invitaciones y tokens de CI.

Toda ruta bajo `:organizationId` resuelve el rol contra la base, en esa organización. No hay
ninguna que lea un identificador de organización y se fíe: el de la URL es una pregunta, y la
membresía es la respuesta.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, Request

from ..crypto import generate_opaque_token, hash_opaque_token, token_preview
from ..db import iso
from ..domain import (
    INVITATION_TTL,
    normalize_email,
    slugify_organization,
    utc_now,
    would_orphan_organization,
)
from ..identity import Principal, authenticate, require_user, role_in_organization
from ..problems import ConflictError, ForbiddenError, NotFoundError
from ..repositories import at_least, new_id
from ..validation import ROLES, Rules
from .auth import StrictModel

router = APIRouter(tags=["iam"])

# El rol llega como `str` y no como `Literal` a propósito: un `Literal` lo rechazaría Pydantic con
# su propio mensaje («Input should be 'viewer', ...») y el front pinta ese texto bajo el campo. La
# comprobación va en `Rules`, con la frase del original.
class CreateOrganizationBody(StrictModel):
    name: str


class InviteMemberBody(StrictModel):
    email: str
    role: str


class ChangeRoleBody(StrictModel):
    role: str


class AcceptInvitationBody(StrictModel):
    token: str


async def create_organization(request: Request, name: str, owner_id: str) -> str:
    """Funda una organización y mete dentro a su propietario.

    La llama también el registro, y por eso es una función y no solo una ruta: una cuenta sin
    organización no puede hacer nada, así que crearla es parte de crear la cuenta y no un paso
    que el cliente tenga que acordarse de dar."""
    state = request.app.state.eq
    now = utc_now()
    base = slugify_organization(name)
    slug = base
    suffix = 2
    while await state.repositories.organizations.find_by_slug(slug):
        slug = f"{base}-{suffix}"
        suffix += 1
        if suffix >= 1000:
            slug = f"{base}-{new_id()[:8]}"
            break

    organization_id = new_id()
    await state.repositories.organizations.insert(
        {"id": organization_id, "name": name.strip(), "slug": slug, "createdAt": now}
    )
    await state.repositories.memberships.save(
        {"organizationId": organization_id, "userId": owner_id, "role": "owner", "createdAt": now}
    )
    return organization_id


@router.post("/orgs", status_code=201)
async def create(request: Request, body: CreateOrganizationBody, principal: Principal = Depends(authenticate)) -> dict:
    Rules().max_length("name", body.name, 200).check()
    organization_id = await create_organization(request, body.name, require_user(principal))
    state = request.app.state.eq
    organization = await state.repositories.organizations.find_by_id(organization_id)
    return {"organizationId": organization_id, "slug": organization["slug"] if organization else ""}


@router.post("/invitations/accept", status_code=200)
async def accept(request: Request, body: AcceptInvitationBody, principal: Principal = Depends(authenticate)) -> dict:
    Rules().max_length("token", body.token, 200).check()
    state = request.app.state.eq
    user_id = require_user(principal)
    now = utc_now()

    invitation = await state.repositories.invitations.find_by_hash(hash_opaque_token(body.token))
    if not invitation or invitation["revokedAt"] is not None or invitation["acceptedAt"] is not None:
        raise NotFoundError("La invitación no es válida", "invitation-invalid")
    if invitation["expiresAt"] <= now:
        raise NotFoundError("La invitación ha caducado", "invitation-expired")

    user = await state.repositories.users.find_by_id(user_id)
    if not user:
        raise NotFoundError("El usuario no existe", "user-not-found")
    if normalize_email(user["email"]) != invitation["email"]:
        raise ForbiddenError("Esta invitación es para otra dirección de correo", "invitation-wrong-recipient")

    organization_id = str(invitation["organizationId"])
    if not await state.repositories.memberships.find(organization_id, user_id):
        await state.repositories.memberships.save(
            {
                "organizationId": organization_id,
                "userId": user_id,
                "role": invitation["role"],
                "createdAt": now,
            }
        )
    await state.repositories.invitations.mark_accepted(str(invitation["id"]), now)
    return {"organizationId": organization_id}


@router.get("/orgs/{organization_id}/members")
async def members(request: Request, organization_id: str, principal: Principal = Depends(authenticate)) -> dict:
    await role_in_organization(request, principal, organization_id, "viewer")
    state = request.app.state.eq

    rows = await state.repositories.memberships.list_for_organization(organization_id)
    people = []
    for membership in rows:
        user = await state.repositories.users.find_by_id(str(membership["userId"]))
        if user:
            people.append(
                {
                    "userId": str(user["id"]),
                    "email": user["email"],
                    "name": user["name"],
                    "role": membership["role"],
                    "since": iso(membership["createdAt"]),
                }
            )

    # El hash del token no sale del repositorio: una invitación pendiente es una credencial viva
    # hasta que se acepta, y esta vista es lo que pinta la pantalla de miembros.
    pending = [
        {
            "id": str(invitation["id"]),
            "email": invitation["email"],
            "role": invitation["role"],
            "expiresAt": iso(invitation["expiresAt"]),
        }
        for invitation in await state.repositories.invitations.list_for_organization(organization_id)
        if invitation["acceptedAt"] is None and invitation["revokedAt"] is None
    ]
    return {"members": people, "invitations": pending}


@router.post("/orgs/{organization_id}/invitations", status_code=201)
async def invite(
    request: Request,
    organization_id: str,
    body: InviteMemberBody,
    principal: Principal = Depends(authenticate),
) -> dict:
    await role_in_organization(request, principal, organization_id, "admin")
    # Después del rol y no antes: los guards corren antes que el pipe en el original, así que
    # quien no es admin recibe 403 aunque además mande un cuerpo inválido.
    Rules().email("email", body.email).max_length("email", body.email, 320).one_of(
        "role", body.role, ROLES
    ).check()
    state = request.app.state.eq
    actor_id = require_user(principal)

    inviter = await state.repositories.memberships.find(organization_id, actor_id)
    if not inviter or not at_least(str(inviter["role"]), "admin"):
        raise ForbiddenError("No puedes invitar a esta organización")
    if not at_least(str(inviter["role"]), body.role):
        raise ForbiddenError("No puedes invitar con un rol superior al tuyo", "role-escalation")

    email = normalize_email(body.email)
    if await state.repositories.invitations.find_pending(organization_id, email):
        raise ConflictError("Esa dirección ya tiene una invitación pendiente", "invitation-pending")

    now = utc_now()
    token = generate_opaque_token()
    invitation_id = new_id()
    expires_at = now + INVITATION_TTL
    await state.repositories.invitations.insert(
        {
            "id": invitation_id,
            "organizationId": organization_id,
            "email": email,
            "role": body.role,
            "tokenHash": hash_opaque_token(token),
            "invitedBy": actor_id,
            "createdAt": now,
            "expiresAt": expires_at,
        }
    )
    # El token se devuelve una vez, como allí: es lo que permite a la pantalla enseñar el enlace
    # cuando no hay correo saliente configurado.
    return {"invitationId": invitation_id, "token": token, "expiresAt": iso(expires_at)}


@router.patch("/orgs/{organization_id}/members/{user_id}", status_code=204)
async def change_role(
    request: Request,
    organization_id: str,
    user_id: str,
    body: ChangeRoleBody,
    principal: Principal = Depends(authenticate),
) -> None:
    await role_in_organization(request, principal, organization_id, "admin")
    Rules().one_of("role", body.role, ROLES).check()
    state = request.app.state.eq
    actor_id = require_user(principal)

    actor = await state.repositories.memberships.find(organization_id, actor_id)
    if not actor or not at_least(str(actor["role"]), "admin"):
        raise ForbiddenError("No puedes gestionar miembros de esta organización")
    if actor_id == user_id:
        raise ForbiddenError("No puedes cambiar tu propio rol", "self-role-change")
    if not at_least(str(actor["role"]), body.role):
        raise ForbiddenError("No puedes otorgar un rol superior al tuyo", "role-escalation")

    target = await state.repositories.memberships.find(organization_id, user_id)
    if not target:
        raise NotFoundError("Esa persona no es miembro de la organización", "membership-not-found")
    # Un admin tampoco degrada a un propietario: la escalera tiene que valer en los dos sentidos o
    # el rango de encima del tuyo es solo una etiqueta.
    if not at_least(str(actor["role"]), str(target["role"])):
        raise ForbiddenError("No puedes modificar a alguien con un rol superior al tuyo", "role-escalation")

    everyone = await state.repositories.memberships.list_for_organization(organization_id)
    if would_orphan_organization(everyone, user_id, body.role):
        raise ConflictError("La organización quedaría sin propietario", "last-owner")

    await state.repositories.memberships.save(
        {
            "organizationId": organization_id,
            "userId": user_id,
            "role": body.role,
            "createdAt": target["createdAt"],
        }
    )


@router.delete("/orgs/{organization_id}/members/{user_id}", status_code=204)
async def remove_member(
    request: Request,
    organization_id: str,
    user_id: str,
    principal: Principal = Depends(authenticate),
) -> None:
    # `viewer` y no `admin`: salir de una organización a la que te invitaron no puede exigir el
    # permiso de gestionar a la gente que hay dentro.
    await role_in_organization(request, principal, organization_id, "viewer")
    state = request.app.state.eq
    actor_id = require_user(principal)

    actor = await state.repositories.memberships.find(organization_id, actor_id)
    if not actor:
        raise ForbiddenError("No perteneces a esta organización")
    target = await state.repositories.memberships.find(organization_id, user_id)
    if not target:
        raise NotFoundError("Esa persona no es miembro de la organización", "membership-not-found")

    leaving = actor_id == user_id
    if not leaving:
        if not at_least(str(actor["role"]), "admin"):
            raise ForbiddenError("No puedes gestionar miembros de esta organización")
        if not at_least(str(actor["role"]), str(target["role"])):
            raise ForbiddenError("No puedes expulsar a alguien con un rol superior al tuyo", "role-escalation")

    everyone = await state.repositories.memberships.list_for_organization(organization_id)
    if would_orphan_organization(everyone, user_id, None):
        raise ConflictError("La organización quedaría sin propietario", "last-owner")

    await state.repositories.memberships.remove(organization_id, user_id)


class CreateApiTokenBody(StrictModel):
    name: str


@router.get("/orgs/{organization_id}/tokens")
async def list_tokens(request: Request, organization_id: str, principal: Principal = Depends(authenticate)) -> list:
    await role_in_organization(request, principal, organization_id, "admin")
    state = request.app.state.eq
    return [
        {
            "id": str(token["id"]),
            "name": token["name"],
            "preview": token["preview"],
            "createdAt": iso(token["createdAt"]),
            "lastUsedAt": iso(token["lastUsedAt"]),
            "revokedAt": iso(token["revokedAt"]),
        }
        for token in await state.repositories.api_tokens.list_for_organization(organization_id)
    ]


@router.post("/orgs/{organization_id}/tokens", status_code=201)
async def create_token(
    request: Request,
    organization_id: str,
    body: CreateApiTokenBody,
    principal: Principal = Depends(authenticate),
) -> dict:
    await role_in_organization(request, principal, organization_id, "admin")
    Rules().max_length("name", body.name, 120).check()
    state = request.app.state.eq
    # Con prefijo para que un token filtrado sea reconocible en un registro o en un repositorio
    # público, por un escáner de secretos y por quien se lo encuentre.
    token = f"eqt_{generate_opaque_token()}"
    token_id = new_id()
    await state.repositories.api_tokens.insert(
        {
            "id": token_id,
            "organizationId": organization_id,
            "name": body.name.strip() or "Token de CI",
            "tokenHash": hash_opaque_token(token),
            "preview": token_preview(token),
            "createdBy": require_user(principal),
            "createdAt": utc_now(),
        }
    )
    return {"id": token_id, "token": token, "preview": token_preview(token)}


@router.delete("/orgs/{organization_id}/tokens/{token_id}", status_code=204)
async def revoke_token(
    request: Request,
    organization_id: str,
    token_id: str,
    principal: Principal = Depends(authenticate),
) -> None:
    await role_in_organization(request, principal, organization_id, "admin")
    state = request.app.state.eq
    token = await state.repositories.api_tokens.find_by_id(token_id)
    # La comprobación de la organización va dentro del 404 y no al lado: contestar 403 por un token
    # que es de otro confirma que ese identificador existe, que es un oráculo de pertenencia entre
    # inquilinos. Para quien llama, no existe.
    if not token or str(token["organizationId"]) != organization_id:
        raise NotFoundError("El token no existe", "api-token-not-found")
    if token["revokedAt"] is not None:
        return
    await state.repositories.api_tokens.revoke(token_id, utc_now())

File: apps/api-go/internal/api/iam.go
package api

// Organizaciones, miembros, invitaciones y tokens de CI.
//
// Toda ruta bajo `{organizationId}` resuelve el rol contra la base, en esa organización. No hay
// ninguna que lea un identificador de organización y se fíe: el de la URL es una pregunta, y la
// membresía es la respuesta.

import (
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/smalbach/endpoint-quality/apps/api-go/internal/crypto"
	"github.com/smalbach/endpoint-quality/apps/api-go/internal/domain"
	"github.com/smalbach/endpoint-quality/apps/api-go/internal/problems"
	"github.com/smalbach/endpoint-quality/apps/api-go/internal/store"
)

type createOrganizationBody struct {
	Name string `json:"name"`
}

type inviteMemberBody struct {
	Email string `json:"email"`
	Role  string `json:"role"`
}

type changeRoleBody struct {
	Role string `json:"role"`
}

type acceptInvitationBody struct {
	Token string `json:"token"`
}

type createAPITokenBody struct {
	Name string `json:"name"`
}

type organizationCreated struct {
	OrganizationID string `json:"organizationId"`
	Slug           string `json:"slug"`
}

type memberView struct {
	UserID string `json:"userId"`
	Email  string `json:"email"`
	Name   string `json:"name"`
	Role   string `json:"role"`
	Since  string `json:"since"`
}

type pendingInvitation struct {
	ID        string `json:"id"`
	Email     string `json:"email"`
	Role      string `json:"role"`
	ExpiresAt string `json:"expiresAt"`
}

type membersView struct {
	Members     []memberView        `json:"members"`
	Invitations []pendingInvitation `json:"invitations"`
}

type invitationCreated struct {
	InvitationID string `json:"invitationId"`
	Token        string `json:"token"`
	ExpiresAt    string `json:"expiresAt"`
}

type apiTokenView struct {
	ID         string  `json:"id"`
	Name       string  `json:"name"`
	Preview    string  `json:"preview"`
	CreatedAt  string  `json:"createdAt"`
	LastUsedAt *string `json:"lastUsedAt"`
	RevokedAt  *string `json:"revokedAt"`
}

type apiTokenCreated struct {
	ID      string `json:"id"`
	Token   string `json:"token"`
	Preview string `json:"preview"`
}

// foundOrganization funda una organización y mete dentro a su propietario. La llama también el
// registro: una cuenta sin organización no puede hacer nada, así que crearla es parte de crear la
// cuenta y no un paso que el cliente tenga que acordarse de dar.
func (s *Server) foundOrganization(request *http.Request, name, ownerID string) (string, string, error) {
	ctx := request.Context()
	base := domain.SlugifyOrganization(name)
	slug := base
	for suffix := 2; suffix < 1000; suffix++ {
		taken, err := s.store.OrganizationSlugTaken(ctx, slug)
		if err != nil {
			return "", "", err
		}
		if !taken {
			break
		}
		slug = fmt.Sprintf("%s-%d", base, suffix)
	}

	now := time.Now().UTC()
	organization := store.Organization{ID: newID(), Name: strings.TrimSpace(name), Slug: slug, CreatedAt: now}
	if err := s.store.InsertOrganization(ctx, organization); err != nil {
		return "", "", err
	}
	if err := s.store.SaveMembership(ctx, store.Membership{
		OrganizationID: organization.ID, UserID: ownerID, Role: "owner", CreatedAt: now,
	}); err != nil {
		return "", "", err
	}
	return organization.ID, slug, nil
}

func (s *Server) createOrganization(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	userID, err := requireUser(who)
	if err != nil {
		return err
	}
	var body createOrganizationBody
	if err := decode(request, &body); err != nil {
		return err
	}
	if err := (rules{}).maxLength("name", body.Name, 200).check(); err != nil {
		return err
	}
	organizationID, slug, err := s.foundOrganization(request, body.Name, userID)
	if err != nil {
		return err
	}
	return writeJSON(writer, http.StatusCreated, organizationCreated{OrganizationID: organizationID, Slug: slug})
}

func (s *Server) acceptInvitation(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	userID, err := requireUser(who)
	if err != nil {
		return err
	}
	var body acceptInvitationBody
	if err := decode(request, &body); err != nil {
		return err
	}
	if err := (rules{}).maxLength("token", body.Token, 200).check(); err != nil {
		return err
	}

	ctx := request.Context()
	now := time.Now().UTC()
	invitation, err := s.store.InvitationByHash(ctx, crypto.HashOpaqueToken(body.Token))
	if err != nil {
		return err
	}
	if invitation == nil || invitation.RevokedAt != nil || invitation.AcceptedAt != nil {
		return problems.NotFound("La invitación no es válida", "invitation-invalid")
	}
	if !invitation.ExpiresAt.After(now) {
		return problems.NotFound("La invitación ha caducado", "invitation-expired")
	}
	user, err := s.store.UserByID(ctx, userID)
	if err != nil {
		return err
	}
	if user == nil {
		return problems.NotFound("El usuario no existe", "user-not-found")
	}
	if domain.NormalizeEmail(user.Email) != invitation.Email {
		return problems.Forbidden("Esta invitación es para otra dirección de correo", "invitation-wrong-recipient")
	}

	existing, err := s.store.Membership(ctx, invitation.OrganizationID, userID)
	if err != nil {
		return err
	}
	if existing == nil {
		if err := s.store.SaveMembership(ctx, store.Membership{
			OrganizationID: invitation.OrganizationID, UserID: userID, Role: invitation.Role, CreatedAt: now,
		}); err != nil {
			return err
		}
	}
	if err := s.store.AcceptInvitation(ctx, invitation.ID, now); err != nil {
		return err
	}
	return writeJSON(writer, http.StatusOK, map[string]string{"organizationId": invitation.OrganizationID})
}

func (s *Server) members(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	organizationID := request.PathValue("organizationId")
	if _, err := s.roleIn(request, who, organizationID, "viewer"); err != nil {
		return err
	}

	ctx := request.Context()
	memberships, err := s.store.MembershipsOfOrganization(ctx, organizationID)
	if err != nil {
		return err
	}
	people := []memberView{}
	for _, membership := range memberships {
		user, err := s.store.UserByID(ctx, membership.UserID)
		if err != nil {
			return err
		}
		if user != nil {
			people = append(people, memberView{
				UserID: user.ID, Email: user.Email, Name: user.Name,
				Role: membership.Role, Since: iso(membership.CreatedAt),
			})
		}
	}

	invitations, err := s.store.Invitations(ctx, organizationID)
	if err != nil {
		return err
	}
	// El hash del token no sale del repositorio: una invitación pendiente es una credencial viva
	// hasta que se acepta, y esta vista es lo que pinta la pantalla de miembros.
	pending := []pendingInvitation{}
	for _, invitation := range invitations {
		if invitation.AcceptedAt == nil && invitation.RevokedAt == nil {
			pending = append(pending, pendingInvitation{
				ID: invitation.ID, Email: invitation.Email, Role: invitation.Role,
				ExpiresAt: iso(invitation.ExpiresAt),
			})
		}
	}
	return writeJSON(writer, http.StatusOK, membersView{Members: people, Invitations: pending})
}

func (s *Server) invite(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	organizationID := request.PathValue("organizationId")
	if _, err := s.roleIn(request, who, organizationID, "admin"); err != nil {
		return err
	}
	userID, err := requireUser(who)
	if err != nil {
		return err
	}
	var body inviteMemberBody
	if err := decode(request, &body); err != nil {
		return err
	}
	// Después del rol y no antes: los guards corren antes que el pipe en el original, así que
	// quien no es admin recibe 403 aunque además mande un cuerpo inválido.
	if err := (rules{}).email("email", body.Email).maxLength("email", body.Email, 320).
		oneOf("role", body.Role, domain.Roles).check(); err != nil {
		return err
	}

	ctx := request.Context()
	inviter, err := s.store.Membership(ctx, organizationID, userID)
	if err != nil {
		return err
	}
	if inviter == nil || !domain.AtLeast(inviter.Role, "admin") {
		return problems.Forbidden("No puedes invitar a esta organización", "")
	}
	if !domain.AtLeast(inviter.Role, body.Role) {
		return problems.Forbidden("No puedes invitar con un rol superior al tuyo", "role-escalation")
	}

	email := domain.NormalizeEmail(body.Email)
	existing, err := s.store.PendingInvitation(ctx, organizationID, email)
	if err != nil {
		return err
	}
	if existing != nil {
		return problems.Conflict("Esa dirección ya tiene una invitación pendiente", "invitation-pending")
	}

	token, err := crypto.GenerateOpaqueToken()
	if err != nil {
		return err
	}
	now := time.Now().UTC()
	invitation := store.Invitation{
		ID: newID(), OrganizationID: organizationID, Email: email, Role: body.Role,
		CreatedAt: now, ExpiresAt: now.Add(domain.InvitationTTL),
	}
	if err := s.store.InsertInvitation(ctx, invitation, crypto.HashOpaqueToken(token), userID); err != nil {
		return err
	}
	// El token se devuelve una vez: es lo que permite a la pantalla enseñar el enlace cuando no hay
	// correo saliente configurado.
	return writeJSON(writer, http.StatusCreated, invitationCreated{
		InvitationID: invitation.ID, Token: token, ExpiresAt: iso(invitation.ExpiresAt),
	})
}

func (s *Server) changeRole(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	organizationID := request.PathValue("organizationId")
	targetID := request.PathValue("userId")
	if _, err := s.roleIn(request, who, organizationID, "admin"); err != nil {
		return err
	}
	actorID, err := requireUser(who)
	if err != nil {
		return err
	}
	var body changeRoleBody
	if err := decode(request, &body); err != nil {
		return err
	}
	if err := (rules{}).oneOf("role", body.Role, domain.Roles).check(); err != nil {
		return err
	}

	ctx := request.Context()
	actor, err := s.store.Membership(ctx, organizationID, actorID)
	if err != nil {
		return err
	}
	if actor == nil || !domain.AtLeast(actor.Role, "admin") {
		return problems.Forbidden("No puedes gestionar miembros de esta organización", "")
	}
	if actorID == targetID {
		return problems.Forbidden("No puedes cambiar tu propio rol", "self-role-change")
	}
	if !domain.AtLeast(actor.Role, body.Role) {
		return problems.Forbidden("No puedes otorgar un rol superior al tuyo", "role-escalation")
	}
	target, err := s.store.Membership(ctx, organizationID, targetID)
	if err != nil {
		return err
	}
	if target == nil {
		return problems.NotFound("Esa persona no es miembro de la organización", "membership-not-found")
	}
	// Un admin tampoco degrada a un propietario: la escalera tiene que valer en los dos sentidos o
	// el rango de encima del tuyo es solo una etiqueta.
	if !domain.AtLeast(actor.Role, target.Role) {
		return problems.Forbidden("No puedes modificar a alguien con un rol superior al tuyo", "role-escalation")
	}
	roles, err := s.rolesOf(request, organizationID)
	if err != nil {
		return err
	}
	if domain.WouldOrphanOrganization(roles, targetID, body.Role) {
		return problems.Conflict("La organización quedaría sin propietario", "last-owner")
	}

	if err := s.store.SaveMembership(ctx, store.Membership{
		OrganizationID: organizationID, UserID: targetID, Role: body.Role, CreatedAt: target.CreatedAt,
	}); err != nil {
		return err
	}
	return noContent(writer)
}

func (s *Server) removeMember(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	organizationID := request.PathValue("organizationId")
	targetID := request.PathValue("userId")
	// `viewer` y no `admin`: salir de una organización a la que te invitaron no puede exigir el
	// permiso de gestionar a la gente que hay dentro.
	if _, err := s.roleIn(request, who, organizationID, "viewer"); err != nil {
		return err
	}
	actorID, err := requireUser(who)
	if err != nil {
		return err
	}

	ctx := request.Context()
	actor, err := s.store.Membership(ctx, organizationID, actorID)
	if err != nil {
		return err
	}
	if actor == nil {
		return problems.Forbidden("No perteneces a esta organización", "")
	}
	target, err := s.store.Membership(ctx, organizationID, targetID)
	if err != nil {
		return err
	}
	if target == nil {
		return problems.NotFound("Esa persona no es miembro de la organización", "membership-not-found")
	}
	if actorID != targetID {
		if !domain.AtLeast(actor.Role, "admin") {
			return problems.Forbidden("No puedes gestionar miembros de esta organización", "")
		}
		if !domain.AtLeast(actor.Role, target.Role) {
			return problems.Forbidden("No puedes expulsar a alguien con un rol superior al tuyo", "role-escalation")
		}
	}
	roles, err := s.rolesOf(request, organizationID)
	if err != nil {
		return err
	}
	if domain.WouldOrphanOrganization(roles, targetID, "") {
		return problems.Conflict("La organización quedaría sin propietario", "last-owner")
	}

	if err := s.store.RemoveMembership(ctx, organizationID, targetID); err != nil {
		return err
	}
	return noContent(writer)
}

func (s *Server) rolesOf(request *http.Request, organizationID string) (map[string]string, error) {
	memberships, err := s.store.MembershipsOfOrganization(request.Context(), organizationID)
	if err != nil {
		return nil, err
	}
	roles := map[string]string{}
	for _, membership := range memberships {
		roles[membership.UserID] = membership.Role
	}
	return roles, nil
}

func (s *Server) listAPITokens(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	organizationID := request.PathValue("organizationId")
	if _, err := s.roleIn(request, who, organizationID, "admin"); err != nil {
		return err
	}
	tokens, err := s.store.APITokens(request.Context(), organizationID)
	if err != nil {
		return err
	}
	views := []apiTokenView{}
	for _, token := range tokens {
		views = append(views, apiTokenView{
			ID: token.ID, Name: token.Name, Preview: token.Preview, CreatedAt: iso(token.CreatedAt),
			LastUsedAt: isoPointer(token.LastUsedAt), RevokedAt: isoPointer(token.RevokedAt),
		})
	}
	return writeJSON(writer, http.StatusOK, views)
}

func (s *Server) createAPIToken(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	organizationID := request.PathValue("organizationId")
	if _, err := s.roleIn(request, who, organizationID, "admin"); err != nil {
		return err
	}
	userID, err := requireUser(who)
	if err != nil {
		return err
	}
	var body createAPITokenBody
	if err := decode(request, &body); err != nil {
		return err
	}
	if err := (rules{}).maxLength("name", body.Name, 120).check(); err != nil {
		return err
	}

	// Con prefijo para que un token filtrado sea reconocible en un registro o en un repositorio
	// público, por un escáner de secretos y por quien se lo encuentre.
	opaque, err := crypto.GenerateOpaqueToken()
	if err != nil {
		return err
	}
	token := "eqt_" + opaque
	name := strings.TrimSpace(body.Name)
	if name == "" {
		name = "Token de CI"
	}
	stored := store.APIToken{
		ID: newID(), OrganizationID: organizationID, Name: name,
		Preview: crypto.TokenPreview(token), CreatedAt: time.Now().UTC(),
	}
	if err := s.store.InsertAPIToken(request.Context(), stored, crypto.HashOpaqueToken(token), userID); err != nil {
		return err
	}
	return writeJSON(writer, http.StatusCreated, apiTokenCreated{
		ID: stored.ID, Token: token, Preview: stored.Preview,
	})
}

func (s *Server) revokeAPIToken(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	organizationID := request.PathValue("organizationId")
	if _, err := s.roleIn(request, who, organizationID, "admin"); err != nil {
		return err
	}
	token, err := s.store.APITokenByID(request.Context(), request.PathValue("tokenId"))
	if err != nil {
		return err
	}
	// La comprobación de la organización va dentro del 404 y no al lado: contestar 403 por un
	// token que es de otro confirma que ese identificador existe, que es un oráculo de pertenencia
	// entre inquilinos. Para quien llama, no existe.
	if token == nil || token.OrganizationID != organizationID {
		return problems.NotFound("El token no existe", "api-token-not-found")
	}
	if token.RevokedAt != nil {
		return noContent(writer)
	}
	if err := s.store.RevokeAPIToken(request.Context(), token.ID, time.Now().UTC()); err != nil {
		return err
	}
	return noContent(writer)
}
